0000-modules/detailed-design/loading-files.md

```diff
@@ -0,0 +1,194 @@
+ignore these files and directories:
+
+* The crate root file itself.
+* Any directory with a name which is not a valid Rust identifier.
+* If the crate root is in the `src` subdirectory of the Cargo manifest
+directory, and there is a directory called `src/bin`, cargo will ignore that
+subdirectory.
+
+In short, cargo will include all appropriately named files inside the directory
+which contains the crate root, except that it will ignore the `src/bin`
+directory.
+
+Packages containing multiple crates which wish to use the default module list
+will need to make sure that they do not have multiple crates rooted in the same
+directory, or within a subdirectory of another crate. The most likely
+problematic crates today are those which have both a `src/lib.rs` and a
+`src/main.rs`. We recommend those crates move their binary crate to the
+`src/bin` directory solution.
+
+While gathering the default module list, cargo will determine if any other
+crate is rooted in a directory which would be collected by the default module
+list, and will instead not pass a `--module` list and issue a warning in
+that case, informing users that they need to rearrange their crates or provide
+a list of modules themselves.
+
+(**Note:** These projects will receive a warning, but will not immediately be
+broken, because the `mod` statements they already contain will continue to pick
+up files.)
+
+#### Tests, examples, and benchmarks
+
+Test, example, and benchmark crates follow a different set of rules. If the
+crate is located in the appropriate top-level directory (`tests`, `examples`,
+and so on), no `--module` list will be collected by default. However,
+subdirectories of these directories will be treated as individual binary
+crates: a `main.rs` file will be treated as the root module, and all other
+appropriately named files will be passed as `--module`, using the same
+rules described above.
+
+So if you have an examples directory like this:
+
+```
+examples/
+    foo.rs
+    bar/
+       main.rs
+       baz.rs
+```
+
+This contains two examples, a `foo` example and a `bar` example, and the `bar`
+crate will have `baz.rs` as a submodule.
+
+The reason for this is that today, cargo will treat every file in `tests`,
+`examples`, and `benches` as independent crates, which is a well-motivated
+design. Usually, these are small enough that a single file makes sense.
+However, today, cargo does not make it particularly easy to have tests,
+examples, or benchmarks that are multiple files. This design will create a
+pattern to enable users to do this.
+
+#### Providing your own module list
+
+The target section of the `Cargo.toml` will  gain a new item called `modules`.
+This item is expected to be an array of strings, which are relative paths from
+the cargo manifest directory to files containing Rust source code. If this item
+is specified, cargo will canonicalize these paths and pass them to rustc as the
+`--module` argument when building this target.
```

Comment:
Just to clarify-- if _any_ module list is passed here, Cargo won't do any directory searching, right?